File: FateGrandAutomata/DroidCvPattern.cs
using System.Collections.Generic;
using System.IO;
using CoreAutomata;
using Org.Opencv.Core;
using Org.Opencv.Imgcodecs;
using Org.Opencv.Imgproc;
using Rect = Org.Opencv.Core.Rect;
using Region = CoreAutomata.Region;
using Size = CoreAutomata.Size;
using Stream = System.IO.Stream;

namespace FateGrandAutomata
{
    public class DroidCvPattern : IPattern
    {
        readonly bool _ownsMat = true;

        public DroidCvPattern() : this(new Mat()) { }

        public DroidCvPattern(Mat Mat, bool OwnsMat = true)
        {
            this.Mat = Mat;
            _ownsMat = OwnsMat;
        }

        public DroidCvPattern(Stream Stream)
        {
            var buffer = new byte[Stream.Length];
            using var ms = new MemoryStream(buffer);
            Stream.CopyTo(ms);

            using var raw = new MatOfByte(buffer);

            Mat = Imgcodecs.Imdecode(raw, Imgcodecs.CvLoadImageColor);
        }

        public Mat Mat { get; }

        public void Dispose()
        {
            if (_ownsMat)
            {
                Mat.Release();
            }
        }

        public IPattern Resize(Size Size)
        {
            var result = new Mat();

            Resize(result, Size);

            return new DroidCvPattern(result);
        }

        void Resize(Mat Target, Size Size)
        {
            Imgproc.Resize(Mat, Target, new Org.Opencv.Core.Size(Size.Width, Size.Height), 0, 0, Imgproc.InterArea);
        }

        public void Resize(IPattern Target, Size Size)
        {
            if (Target is DroidCvPattern target)
            {
                Resize(target.Mat, Size);
            }
        }

        public IPattern Crop(Region Region)
        {
            var rect = new Rect(Region.X, Region.Y, Region.W, Region.H);

            if (rect.X + rect.Width > Width)
            {
                rect.X = Width - rect.Width;
            }

            if (rect.Y + rect.Height > Height)
            {
                rect.Y = Height - rect.Height;
            }

            var result = new Mat(Mat, rect);

            return new DroidCvPattern(result);
        }

        public void Save(string Filename)
        {
            Imgcodecs.Imwrite(Filename, Mat);
        }

        public bool IsMatch(IPattern Template, double Similarity)
        {
            using var result = new DisposableMat();

            Imgproc.MatchTemplate(Mat, (Template as DroidCvPattern)?.Mat, result.Mat, Imgproc.TmCcoeffNormed);

            using var minMaxLocResult = Core.MinMaxLoc(result.Mat);

            return minMaxLocResult.MaxVal >= Similarity;
        }

        public int Width => Mat.Width();

        public int Height => Mat.Height();

        public IEnumerable<Match> FindMatches(IPattern Template, double Similarity)
        {
            using var result = new DisposableMat();

            // max is used for tmccoeff
            Imgproc.MatchTemplate(Mat, (Template as DroidCvPattern)?.Mat, result.Mat, Imgproc.TmCcoeffNormed);

            Imgproc.Threshold(result.Mat, result.Mat, 0.1, 1, Imgproc.ThreshTozero);

            while (true)
            {
                using var minMaxLocResult = Core.MinMaxLoc(result.Mat);
                var score = minMaxLocResult.MaxVal;

                if (score >= Similarity)
                {
                    var loc = minMaxLocResult.MaxLoc;
                    var region = new Region((int)loc.X, (int)loc.Y, Template.Width, Template.Height);

                    yield return new Match(region, score);

                    using var mask = new DisposableMat();
                    // Flood fill eliminates the problem of nearby points to a high similarity point also having high similarity
                    const double floodFillDiff = 0.05;
                    Imgproc.FloodFill(result.Mat, mask.Mat, loc, new Scalar(0),
                        new Rect(),
                        new Scalar(floodFillDiff), new Scalar(floodFillDiff),
                        0);
                }
                else break;
            }
        }

        public IPattern Copy()
        {
            return new DroidCvPattern(Mat.Clone());
        }
    }
}

File: app/src/main/java/com/mathewsachin/fategrandautomata/imaging/DroidCvPattern.kt
package com.mathewsachin.fategrandautomata.imaging

import android.graphics.Bitmap
import com.mathewsachin.libautomata.Match
import com.mathewsachin.libautomata.Pattern
import com.mathewsachin.libautomata.Region
import com.mathewsachin.libautomata.Size
import org.opencv.core.*
import org.opencv.imgcodecs.Imgcodecs
import org.opencv.imgproc.Imgproc
import timber.log.Timber
import java.io.InputStream
import java.io.OutputStream
import kotlin.math.roundToInt
import org.opencv.core.Size as CvSize

class DroidCvPattern(
    private var mat: Mat = Mat(),
    private val ownsMat: Boolean = true
) : Pattern {
    private companion object {
        fun makeMat(
            stream: InputStream,
            isColor: Boolean
        ): Mat {
            val byteArray = stream.readBytes()

            return MatOfByte(*byteArray).use {
                Imgcodecs.imdecode(
                    it,
                    if (isColor) Imgcodecs.IMREAD_COLOR else Imgcodecs.IMREAD_GRAYSCALE
                )
            }
        }
    }

    constructor(stream: InputStream, isColor: Boolean) : this(makeMat(stream, isColor))

    private var tag = ""

    override fun toString() =
        tag.ifBlank { super.toString() }

    override fun close() {
        if (ownsMat) {
            mat.release()
        }
    }

    private fun resize(source: Mat, target: Mat, size: Size) {
        Imgproc.resize(
            source, target,
            CvSize(size.width.toDouble(), size.height.toDouble()),
            0.0, 0.0, Imgproc.INTER_AREA
        )
    }

    override fun resize(size: Size): Pattern {
        val resizedMat = Mat().apply { resize(mat, this, size) }

        return DroidCvPattern(resizedMat).tag(tag)
    }

    override fun resize(target: Pattern, size: Size) {
        if (target is DroidCvPattern) {
            resize(mat, target.mat, size)
        }

        target.tag(tag)
    }

    private fun match(template: Pattern): Mat {
        val result = Mat()
        if (template is DroidCvPattern) {
            if (template.width <= width && template.height <= height) {
                Imgproc.matchTemplate(
                    mat,
                    template.mat,
                    result,
                    Imgproc.TM_CCOEFF_NORMED
                )
            } else {
                Timber.v("Skipped matching $template: Region out of bounds")
            }
        }

        return result
    }

    override fun findMatches(template: Pattern, similarity: Double) = sequence {
        val result = match(template)

        result.use {
            while (true) {
                val minMaxLocResult = Core.minMaxLoc(it)
                val score = minMaxLocResult.maxVal

                if (score >= similarity) {
                    val loc = minMaxLocResult.maxLoc
                    val region = Region(
                        loc.x.roundToInt(),
                        loc.y.roundToInt(),
                        template.width,
                        template.height
                    )

                    val match = Match(region, score)

                    Timber.d("Matched $template with a score of ${match.score}")
                    yield(match)

                    Mat().use { mask ->
                        // Flood fill eliminates the problem of nearby points to a high similarity point also having high similarity
                        val floodFillDiff = 0.3
                        Imgproc.floodFill(
                            result, mask, loc, Scalar(0.0),
                            Rect(),
                            Scalar(floodFillDiff), Scalar(floodFillDiff),
                            Imgproc.FLOODFILL_FIXED_RANGE
                        )
                    }
                } else {
                    Timber.v("Stopped matching $template at score ($score) < similarity ($similarity)")
                    break
                }
            }
        }
    }

    override val width get() = mat.width()
    override val height get() = mat.height()

    override fun crop(region: Region): Pattern {
        val clippedRegion = Region(0, 0, width, height)
            .clip(region)

        val rect = Rect(clippedRegion.x, clippedRegion.y, clippedRegion.width, clippedRegion.height)

        return DroidCvPattern(Mat(mat, rect)).tag(tag)
    }

    fun asBitmap(): Bitmap {
        val bmp = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)

        Mat().use {
            val conversion = if (mat.type() == CvType.CV_8UC1)
                Imgproc.COLOR_GRAY2RGB
            else Imgproc.COLOR_BGR2RGBA

            Imgproc.cvtColor(mat, it, conversion)
            org.opencv.android.Utils.matToBitmap(it, bmp)
        }

        return bmp
    }

    override fun save(stream: OutputStream) {
        asBitmap().use { bmp ->
            bmp.compress(Bitmap.CompressFormat.PNG, 90, stream)
        }
    }

    override fun copy() = DroidCvPattern(mat.clone()).tag(tag)

    override fun tag(tag: String) = apply { this.tag = tag }

    override fun threshold(value: Double): Pattern {
        val result = Mat()

        Imgproc.threshold(mat, result, value * 255, 255.0, Imgproc.THRESH_BINARY)

        return DroidCvPattern(result)
            .tag("$tag[threshold=$value]")
    }

    override fun isWhite(): Boolean {
        val minMaxLocResult = Core.minMaxLoc(mat)
        // 0 = black, 255 = white
        return minMaxLocResult.minVal >= 200
    }
}